elastic/eui
# [EuiFlyout] Include all fixed `EuiHeader`s in flyout focus traps + improve screen reader accessibility

## Summary

closes https://github.com/elastic/eui/issues/5206

Fixed `EuiHeader`s that sit next to `EuiFlyout`s and can still be interacted with at the same time as `EuiFlyout`s need to be automatically added to the flyout's `focusTrapProps.shards` API in order for the flyout to not consider the header "outside of the flyout" and trigger focus trap close.

What this PR does:
- Adds that OOTB functionality (primarily targeted at Kibana and Kibana's global sitewide search box). This can be turned off if desired by consumers via the new `includeFixedHeadersInFocusTrap` prop.
- Additionally improves default EuiFlyout accessibility with the following changes:
  - Auto-focuses the flyout wrapper (to match `EuiPopover` and `EuiModal` - as opposed to the close button (or the first focusable item if the close button has been removed)) 
  - Adds detailed screen reader instructions (matching the detail level of existing `EuiPopover` screen reader descriptions)
  - Adds screen reader details about fixed headers now existing in the focus trap tab rotation
  - Removes the ability for consumers to override `role="dialog"`
- Incidentally removes `closeButtonAriaLabel` - it's unnecessary as `closeButtonProps: { 'aria-label': 'something' }` does the same thing

What this PR **DOES NOT** do:
- Address existing accessibility/focus issues with `push` type EuiFlyouts
- Address existing accessibility details of `ownFocus={false}`
- The above a11y issues have been logged separately in #6576

## QA

The focus fighting issue was already afflicting our ["Elastic navigation pattern" demo](https://elastic.github.io/eui/#/layout/header/elastic-pattern) on prod. This PR can be compared directly against prod behavior by opening a flyout or nav and then clicking the sitewide search input.

- Go to https://eui.elastic.co/pr_6566/#/layout/header/elastic-pattern
- Open either the collapsible nav (leftmost hamburger menu)
- [x] Confirm that screen reader instructions are read out describing `Main navigation, web dialog`
- Once the nav is open, click the "Search for anything" global search in the top nav
- [x] Confirm that the sitewide search popover stays open and works as expected
- [x] Confirm that you can still tab through all headers and the flyout
- [x] Close the nav and confirm that focus is restored to the correct toggling button
- Close the collapsible nav
- Open the alerts flyout by clicking the party icon with the pink dot in the top right
- [x] Confirm that screen reader instructions are read out that include information about the header
- Once the flyout is open, click the "Search for anything" global search in the top nav
- [x] Confirm that the sitewide search popover stays open and works as expected
- [x] Confirm that you can still tab through all headers and the flyout
- [x] Close the nav and confirm that focus is restored to the correct toggling button

### General checklist

- [x] Props have proper **autodocs** (using `@default` if default values are missing) and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**

~- [ ] Checked in both **light and dark** modes~
~- [ ] Checked in **mobile**~
~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**~
~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
~- [ ] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart~
